src/inmanta/parser/plyInmantaParser.py

```diff
@@ -693,6 +694,41 @@ def p_boolean_expression_is_defined_short(p: YaccProduction) -> None:
     attach_lnr(p)
 
 
+"""
+Solution 1
+- Both (1) and (2) work as intended on their own but I can't tie them up in an "And" node
+- I don't see how to check that a key belongs to the dict with this solution
+"""
+# def p_boolean_expression_is_defined_map_lookup(p: YaccProduction) -> None:
+#     """boolean_expression : map_lookup IS DEFINED %prec IS_DEFINED"""
+#
+#     # p[0] = NotEqual(p[1], Literal(list()))                                            # (1)
+#     # p[0] = NotEqual(p[1], Literal(NoneValue()))                                       # (2)
+#     p[0] = And(NotEqual(p[1], Literal(list())), NotEqual(p[1], Literal(NoneValue())))   # And(1, 2)
+#     attach_lnr(p, 2)
+
+"""
+Solution 2
+- easier to check for key belonging to dict
+- (1), (2) and (3) work as intended on their own but again I can't tie it up in an "And" node
+- only base case with var_ref considered here (attributes and nested dicts need to be added)
+"""
+
+
+def p_boolean_expression_is_defined_not_map_lookup(p: YaccProduction) -> None:
+    """boolean_expression : var_ref '[' operand ']' IS DEFINED %prec IS_DEFINED"""
+
+    # | attr_ref '[' operand ']' IS DEFINED %prec IS_DEFINED
+    # | map_lookup '[' operand ']' IS DEFINED %prec IS_DEFINED
+
+    key_in_dict = In(p[3], p[1])  # (1)
+    not_none = NotEqual(MapLookup(p[1], p[3]), Literal(NoneValue()))  # (2)
+    not_empty_list = NotEqual(MapLookup(p[1], p[3]), Literal(list()))  # (3)
```

Comment:
Our general list constructor is `CreateList` rather than `Literal` but I think this should work regardless. I don't immediately see why this rule does not work. I'll need to have a look.